src/main/groovy/com/twcable/grabbit/spring/batch/repository/JcrGrabbitJobInstanceDao.groovy

```diff
@@ -257,10 +257,21 @@ class JcrGrabbitJobInstanceDao extends AbstractJcrDao implements GrabbitJobInsta
         if (!resolver) throw new IllegalArgumentException("resolver == null")
 
         final rootResource = resolver.getResource(JOB_INSTANCE_ROOT)
-        final nextId = (rootResource.children.asList().size() + 1) as Long
+
+        final lastInstance = rootResource?.children?.asList()?.max { Resource resource ->
+            final properties = resource.adaptTo(ValueMap)
+            properties[INSTANCE_ID] as Long
+        }
+        final lastInstanceProperties = lastInstance?.adaptTo(ValueMap)
+        Long nextId = lastInstanceProperties?.get(INSTANCE_ID) as Long
+        if (!nextId) {
+            nextId = 1
+        } else {
+            nextId += 1
+        }
```

Comment:
There appears to be a concurrency issue.

It's reading a value from the JCR and incrementing it, but it's not transactionally updating the JCR. So if another thread tried to ask for the next id at "the same time" (in however many milliseconds/seconds it took to write the update) there would be a race condition.

(Note the existing code has the same problem, just not _quite_ as pronounced.)